src/Microsoft.PowerShell.Commands.Management/commands/management/TestConnectionCommand.cs

```diff
@@ -275,53 +286,109 @@ protected override void ProcessRecord()
-                    Task.WhenAny(connectionTask, timeoutTask).Result.Wait();
+                    stopwatch.Start();
 
-                    if (timeoutTask.Status == TaskStatus.Faulted || timeoutTask.Status == TaskStatus.Canceled)
+                    if (client.ConnectAsync(targetAddress, TcpPort).Wait(TimeoutSeconds * 1000, cancellationTokenSource.Token))
                     {
-                        // Waiting is interrupted by Ctrl-C.
-                        WriteObject(false);
-                        return;
+                        latency = stopwatch.ElapsedMilliseconds;
+                        status = SocketError.Success;
+                    }
+                    else
+                    {
+                        status = SocketError.TimedOut;
                     }
+                }
+                catch (AggregateException ae)
+                {
+                    ae.Handle((ex) =>
+                    {
+                        if (ex is TaskCanceledException)
+                        {
+                            throw new PipelineStoppedException();
+                        }
+                        if (ex is SocketException)
+                        {
+                            SocketException? socketException = ex as SocketException;
+                            status = socketException!.SocketErrorCode;
+                            return true;
+                        }
+                        else
+                        {
+                            return false;
+                        }
+                    });
+                }
+                finally
+                {
+                    stopwatch.Reset();
+                }
 
-                    if (connectionTask.Status == TaskStatus.RanToCompletion)
+                if (Quiet.IsPresent)
+                {
+                    if (status == SocketError.Success)
                     {
                         WriteObject(true);
                         return;
                     }
+                    else if (i == Count)
+                    {
+                        WriteObject(false);
+                        return;
+                    }
+                }
+                else
+                {
+                    WriteObject(new TcpPortStatus(
+                        i,
+                        Source,
+                        targetNameOrAddress,
+                        targetAddress,
+                        TcpPort,
+                        latency,
+                        status == SocketError.Success,
+                        status
+                    ));
                 }
-            }
-            catch
-            {
-                // Silently ignore connection errors.
-            }
-            finally
-            {
-                client.Close();
-            }
 
-            WriteObject(false);
+                Task.Delay(new TimeSpan(0, 0, i == Count ? 0 : Delay)).Wait(cancellationTokenSource.Token);
```

Comment:
Maybe use do-while to avoid last extra delay instead of the trick with ` i == Count ? 0 : Delay`?
```c#
int i = 0;
do
{

}
while ++i < Count && Task.Delay(new TimeSpan(0, 0, Delay)).Wait(cancellationTokenSource.Token);

Reply:
Good point! After looking at this with fresh eyes, the ternary is a bit poor for readability 😅

I found the `do/while` was a bit confusing also in this case, and a straightforward `if` may be the most readable. What do you think?

```csharp
if (i < Count)
{
    Task.Delay(new TimeSpan(0, 0, Delay)).Wait(cancellationTokenSource.Token);
}
```

Reply:
We already use the do-while pattern in the file.

Reply:
I see what you mean, I've checked out the other do-while instances in that file, and not sure if it applies totally here. 

In the case that we use the do-while, aren't we still going to need to implement the `if` statement around the timeout delay? In your example, I see that set up as a nice one-liner, however I'm not able to get that working on my end. I'm running into the following, and am not able to get around it: 
```
Operator '&&' cannot be applied to operands of type 'bool' and 'void'
```
Please let me know if I'm missing something!

Reply:
It is a bug in the code - Task.Wait() returns bool only if millisecondsTimeout parameter presents.

Reply:
I see what you mean, this definitely sounds like a bug! 

How do you reckon we work around this for now?

Reply:
You can use your proposal. :-)

Reply:
Sounds like a plan! I'll make that change asap :)